service/src/main/kotlin/fi/espoo/evaka/varda/VardaUpdateServiceV2.kt

```diff
@@ -0,0 +1,698 @@
+        - For each modified service need delete old related data from varda and add new
+    2. Find out all changed evaka fee data affecting service needs not yet updated above, and for each service need
+       update all service need related data to varda
+ */
+fun updateChildData(db: Database.Connection, client: VardaClient, startingFrom: HelsinkiDateTime) {
+    retryUnsuccessfulServiceNeedVardaUpdates(db, client)
+
+    val serviceNeedDiffsByChild = calculateEvakaVsVardaServiceNeedChangesByChild(db, startingFrom)
+    val feeAndVoucherDiffsByServiceNeed = db.read { it.getServiceNeedFeeData(startingFrom, null) }.groupBy { it.serviceNeedId }
+    val processedServiceNeedIds = mutableSetOf<UUID>()
+
+    serviceNeedDiffsByChild.entries.forEach { serviceNeedDiffByChild ->
+        serviceNeedDiffByChild.value.deletes.forEach { deletedServiceNeedId ->
+            processedServiceNeedIds.add(deletedServiceNeedId)
+            handleDeletedEvakaServiceNeed(db, client, deletedServiceNeedId)
+        }
+
+        serviceNeedDiffByChild.value.updates.forEach { updatedServiceNeedId ->
+            processedServiceNeedIds.add(updatedServiceNeedId)
+            handleUpdatedEvakaServiceNeed(db, client, updatedServiceNeedId)
+        }
+
+        serviceNeedDiffByChild.value.additions.forEach { addedServiceNeedId ->
+            processedServiceNeedIds.add(addedServiceNeedId)
+            handleNewEvakaServiceNeed(db, client, addedServiceNeedId)
+        }
+    }
+
+    // Handle fee data only -changes (no changes in service need). If there was any service need change, all related
+    // fee data has already been sent to / deleted from varda, so we filter those service needs out here
+    feeAndVoucherDiffsByServiceNeed.filterNot { processedServiceNeedIds.contains(it.key) }.entries.forEach {
+        val serviceNeedId = it.key
+        val vardaServiceNeed = db.read { it.getVardaServiceNeed(serviceNeedId) }
+        if (vardaServiceNeed != null) {
+            handleUpdatedEvakaServiceNeed(db, client, serviceNeedId)
+        } else {
+            handleNewEvakaServiceNeed(db, client, serviceNeedId)
+        }
+    }
+}
+
+fun handleDeletedEvakaServiceNeed(db: Database.Connection, client: VardaClient, serviceNeedId: UUID) {
+    try {
+        val vardaServiceNeed = db.read { it.getVardaServiceNeed(serviceNeedId) }
+        if (vardaServiceNeed != null) {
+            val errors = deleteServiceNeedDataFromVarda(client, vardaServiceNeed)
+            if (errors.isNotEmpty()) {
+                logger.error(errors.joinToString(","))
+                db.transaction { it.markServiceNeedUpdateFailed(serviceNeedId, errors) }
+            } else {
+                db.transaction { it.deleteVardaServiceNeedByEvakaServiceNeed(serviceNeedId) }
+            }
+        }
+    } catch (e: Exception) {
+        logger.error("VardaUpdate: manual check needed: something went from while trying to delete varda service need $serviceNeedId data: ${e.localizedMessage}")
```

Comment:
Just a trivial typo here: from -> wrong. Same appears in the next two error messages.